### DISCUSSION TITLE
Does ImageBuilder support Compiling in some specified directories.

### DISCUSSION THREAD

**Discussion by raochp** [o3de/o3de#9110] [Help Wanted]
### Does ImageBuilder support Compiling in some specified directories?
**Hi，I was recently debugging the AssetBuilder feature on O3DE. Then I found some problems.**
I find that I can't builder imge  in some specified directories. When I start AssetBuilder,  It will scan all directories for finding modified image. These make me feel bad.

**finally,expecting a reply. thank you very much.**

I have learn image builder form the following code:

`void BuilderPluginComponent::Activate()
    {
        //load qt plugins for some image file formats support
        AzQtComponents::PrepareQtPaths();

        // create and initialize BuilderSettingManager once since it's will be used for image conversion
        BuilderSettingManager::CreateInstance();

        auto outcome = BuilderSettingManager::Instance()->LoadConfig();
        AZ_Error("Image Processing", outcome.IsSuccess(), "Failed to load Atom image builder settings.");

        if (!outcome.IsSuccess())
        {
            return;
        }

        // Activate is where you'd perform registration with other objects and systems.
        // Since we want to register our builder, we do that here:
        AssetBuilderSDK::AssetBuilderDesc builderDescriptor;
        builderDescriptor.m_name = "Atom Image Builder";for (int i = 0; i < s_TotalSupportedImageExtensions; i++)
        {
            builderDescriptor.m_patterns.push_back(AssetBuilderSDK::AssetBuilderPattern(s_SupportedImageExtensions[i], AssetBuilderSDK::AssetBuilderPattern::PatternType::Wildcard));
        }

        builderDescriptor.m_busId = azrtti_typeid<ImageBuilderWorker>();
        builderDescriptor.m_createJobFunction = AZStd::bind(&ImageBuilderWorker::CreateJobs, &m_imageBuilder, AZStd::placeholders::_1, AZStd::placeholders::_2);
        builderDescriptor.m_processJobFunction = AZStd::bind(&ImageBuilderWorker::ProcessJob, &m_imageBuilder, AZStd::placeholders::_1, AZStd::placeholders::_2);
        builderDescriptor.m_version = 27;   // [ATOM-16958]
        builderDescriptor.m_analysisFingerprint = ImageProcessingAtom::BuilderSettingManager::Instance()->GetAnalysisFingerprint();
        m_imageBuilder.BusConnect(builderDescriptor.m_busId);
        AssetBuilderSDK::AssetBuilderBus::Broadcast(&AssetBuilderSDK::AssetBuilderBusTraits::RegisterBuilderInformation, builderDescriptor);

        m_assetHandlers.emplace_back(AZ::RPI::MakeAssetHandler<AZ::RPI::ImageMipChainAssetHandler>());
        m_assetHandlers.emplace_back(AZ::RPI::MakeAssetHandler<AZ::RPI::StreamingImageAssetHandler>());

        ImageProcessingRequestBus::Handler::BusConnect();
        ImageBuilderRequestBus::Handler::BusConnect();
    }`

**Comment by AMZN-stankowi** [comment]
Hi, the two three concepts to understand here are [Scan Directories](https://development--o3deorg.netlify.app/docs/user-guide/assets/pipeline/scan-directories/), [Source Assets](https://development--o3deorg.netlify.app/docs/user-guide/assets/pipeline/source-assets/), and [Asset Builders](https://development--o3deorg.netlify.app/docs/user-guide/assets/pipeline/asset-builders/)

Scan Directories are the locations that the Asset Processor will look for Source Assets to process. You've found the Builder Descriptor for the Image Builder, this is where the Builder that processes Source Assets matching a certain pattern is registered.

The basic flow is:

- At launch of the Asset Processor, it loads your project and engine configuration to understand which Scan Directories it is monitoring.
- Then, it loads all Asset Builders, and for each file it finds in a Scan Directory, it checks if there is an Asset Builder with a registered pattern that matches. 
    - In the code you've pasted above, this is defined for images by the line that starts with "builderDescriptor.m_patterns.push_back", that's appending all file extensions the image builder is registered again.
- If a file matches one or more builder descriptor patterns, then it's a candidate to become a Source Asset.
   - Multiple builders can process the same source asset and create different product assets.
- After the Asset Processor has collected all of these files, and matched them against the registered Asset Builder descriptor patterns, it then calls CreateJobs for that builder for each of those files.
- This is where the Asset Builder checks the given file and decides if it actually needs to do work for this file. In most cases it will create a job and continue, but sometimes the builder may have additional checks before doing so.
- The Asset Processor next calls ProcessJobs based on the output from CreateJobs. This is where each Asset Builder does the work to create the Product Asset for the Source Asset it is given.

So for your specific situation, the end result is the Asset Processor will attempt to process all image files placed in valid Scan Directories for your project. If you don't want an image processed, then you'll need to save it outside a Scan Directory. If an image isn't processing and you want it to process, then you'll need to save it in a Scan Directory for your project. For most teams, you'll use your project's asset folder and any subfolders in it as the Scan Directory for most content.

I'll move this to a Github Discussion, let me know if you have any follow up questions.

**Reply by raochp** [reply]
**@AMZN-stankowi 
Thank you for your reply. It's very important to me.
About your  reply, I have a  question that I don't understand.
 According to you, I can set my project's asset folder and any subfolders   as the Scan Directory. But  Asset Processor scans all the directories(include Gems,Engine & Editor) actually as shown in the following figure.**  

![20220424-102417](https://user-images.githubusercontent.com/51946066/164953502-1a26e672-413c-447d-899a-f0da8b656eb7.png)

![image](https://user-images.githubusercontent.com/51946066/164952626-1fff1642-0e18-4ea0-8ded-ce1ae5a9cefb.png)
from https://development--o3deorg.netlify.app/docs/user-guide/assets/pipeline/scan-directories/

**I think I didn't set up the correct scan path.  I hope you can tell me how to set the project path or specific path  as a scan path.
I'm looking forward to your help with this question.**
